File: public/dsa/two_pointers/two_sum_sorted/optimal/two_sum_sorted.cpp
#include <bits/stdc++.h>
using namespace std;

bool hasTwoSumZero(vector<int> A) {
    int n = A.size();
    int left = 0, right = n - 1;
    while (left < right) {
        if (A[left] + A[right] == 0) {
            return true;
        } else if (A[left] + A[right] < 0) {
            left++;
        } else {
            right--;
        }
    }
    return false;
}

// tests-start

int main() {
    vector<int> A = {-3, 1, 3, 4};
    vector<int> B = {-2, 1, 3, 4};
    cout << boolalpha << hasTwoSumZero(A) << endl;
    cout << boolalpha << hasTwoSumZero(B) << endl;
    return 0;
}

// tests-end

/*output
true
false
*/

File: public/dsa/two_pointers/two_sum_sorted/optimal/two_sum_sorted.py
from typing import List


class Solution:
    def hasTwoSumZero(self, A: List[int]) -> bool:
        n = len(A)
        left = 0
        right = n - 1
        while left < right:
            if A[left] + A[right] == 0:
                return True
            elif A[left] + A[right] < 0:
                # we need a bigger number
                left += 1
            else:
                # A[left] + A[right] > 0
                # we need a smaller number
                right -= 1
        return False


# tests-start
sol = Solution()
A = [-3, 1, 3, 4]
print(sol.hasTwoSumZero(A))
B = [-2, 1, 3, 4]
print(sol.hasTwoSumZero(B))
# tests-end

'''output
True
False
'''
